### PR TITLE
pass whole array to strategies and add clone

### LINKED ISSUE #125: [BUG] - Using strategies on Array types runs the strategy on array elements instead of the whole array
**Describe the bug** 
When serializing there is a check for `Array.isArray(value)` which, if true, runs the serialization on the individual elements of the array and returns an array of the serialized objects. This is fine in most cases but if you want a custom strategy to serialize an array into a string for example, the behaviour doesnt work, since now you will get an array of the serialized strings. When using strategies at least, the user should be allowed to define the strategy to run on the whole object no matter if it is an array or not. 

**To Reproduce**
```
// Lets say we have an object with a property that is an array of objects that we want to serialize into one single string
@SerializeProperty({ toJSONStrategy: objArraySerialize, fromJSONStrategy: objArrayDeserialize })
public myProp:{id: number}[];
...

// We expect the strategy will be called with the full array and we can serialize it in one shot. 
const objArraySerialize = (items: {id: number}[]): string {
  return JSON.stringify(items.map(el => el.id));
}

```

In the above what will actually happen is that our `objArraySerialize` will get called for every element in the array instead of the full array.

**Expected behaviour** 
The user should be able to define custom strategies for arrays that use the custom strategy for the entire array rather than just the elements on the array

### LINKED ISSUE #132: Add clone functionality
Allow authors to clone an object, should return a new object deep copied

### PR DESCRIPTION
**Pull request by hardy925**
**Description**

pass arrays to strategies, fixes #125, fixes #132

**Things to look at**

- [x] Test coverage
- [x] Code Style
- [x] Documentation (`README.md`, `doc/`, `examples/`, etc..)

### CODE REVIEW

File: strategy/to_json/from_serializable_test.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      serializeMe = 999;
+    }
+    class Test2 extends Serializable {
+      @FromSerializable("outer_property")
+      nested: Test1[] = [new Test1()];
+    }
+
+    class Test3 extends Serializable {
+      @FromSerializable("outer_outer_property")
+      nested2: Test2[] = [new Test2()];
```
Comment (ms1111) [inline strategy/to_json/from_serializable_test.ts:37]: Is the `@FromSerializable` needed here? I would think if we had
```
@SerializeProperty("outer_outer_property")
nested2: Test2[] = [new Test2()];
```
... then the default strategy would map through the array and call tsSerialize() if it's serializable.

File: serializable.ts [hunk 2]
```diff
@@ -148,16 +148,7 @@ export function toPojo(
...
-      record[serializedKey] = value.map((item: JSONValue) => {
-        if (item instanceof Serializable) {
-          return toPojo(item);
-        }
-        return toJSONStrategy(item);
-      });
-    } else if (value !== undefined) {
-      record[serializedKey] = toJSONStrategy(value);
-    }
+    record[serializedKey] = toJSONStrategy(value);
```
Comment (ms1111) [inline serializable.ts]: I think there's an important change to the default behaviour for serialization here. Previously, if you had an array property and were using the default strategy:

```
@SerializeProperty()
private someArray: SomeClass[];
```
... we'd automatically call toPojo() on each element in the array.

Now the default is that we'll call toJSONStrategy(), which just returns the original value. So serialization that would have worked properly before by default, no longer will.

Granted, you always had to give us more information for _deserialization_ to work, using `toSerializable()` in a `fromJSONStrategy`. But previously you could get away without a `toJSONStrategy` and it would do the right thing, kind of by accident, if you only cared about that direction. So that might be worth updating in the examples.

For our own use, we use a shortcut decorator `@DeserializeAs()`:

```ts
export function DeserializeAs<T>(type: SerializableConstructor<T>, propertyName?: string): PropertyDecorator {
    const objectSerializer: SerializePropertyArgument = {
        fromJSONStrategy: toSerializable(type),
    };
    if (propertyName) {
        objectSerializer.serializedKey = propertyName;
    }

    return SerializeProperty(objectSerializer);
}
```
... we would just need to change it to specify a toJsonStrategy of fromSerializable(), so it's not a big change.

**Review by ms1111 — DISMISSED** [review]

File: strategy/to_json/from_serializable_test.ts (hunk 1 above)
Comment (hardy925) [reply strategy/to_json/from_serializable_test.ts:37]: ah so this change runs against the whole array now and not per item, because under the hood `toJSON()` is called and returns the string, so we need to manually say to use `tsSerialize`. This is actually done on purpose because JSON is a text format, and we are trying hard to follow JS API's. for example 

```ts
new Date().toJSON() // returns string value and not a POJO
```

File: serializable.ts (hunk 2 above)
Comment (hardy925) [reply serializable.ts]: Yes 💯 

This is a breaking change and will not be released until v2 is ready. There are a few other features for us to take care of and I only recently and started to set up and plan v2. I am going to add `new Example().clone(arg: Partial<Example>)` or something similar, and we are going to start supporting common decorators, although I'd like to discuss how that will look before starting.

Comment (hardy925) [reply serializable.ts]: Another breaking change is with `.fromJSON()` it only takes a `JSONValue` now, IDK if this is the best approach but I currently believe it is better than using `Object` to support angular and rather have integrators use typing with angular.

Again that is also up for discussion.

Comment (hardy925) [reply serializable.ts]: oh damn! I just noticed I am bringing in a bug by allowing `undefined` to be processed by strategies. 

Thanks, @ms1111

File: serializable.ts [hunk 3]
```diff
@@ -103,6 +107,15 @@ export abstract class Serializable {
   public tsSerialize(): JSONObject {
     return toPojo(this);
   }
+  public clone(jsonObject: Partial<this> = {}): this {
```
Comment (ChrisDufourMB) [inline serializable.ts:112]: I feel that we need a JSDoc comment here explaining what the argument does. It may not be obvious to users who use ts_serialize differently than we do.

File: serializable.ts [hunk 4]
```diff
@@ -40,6 +40,11 @@ export declare interface Serialize {
   tsSerialize(): JSONObject;
 }
 
+/** deep copy `this` */
+export declare interface Clone {
+  clone(jsonObject: Partial<this>): this;
+}
+
```
Comment (ChrisDufourMB) [inline serializable.ts]: Is this interface being used anywhere? Is this different from the method on `Serializable`?

**Review by ChrisDufourMB — DISMISSED** [review]
Changes LGTM, have a few comments surrounding the clone functionality

File: serializable.ts (hunk 4 above)
Comment (hardy925) [reply serializable.ts]: I made these interfaces for implementing if authors wanted to, thinking about it now though I am unsure if we need clone...

File: serializable.ts (hunk 3 above)
Comment (hardy925) [reply serializable.ts:112]: added on the interrface

File: serializable.ts (hunk 4 above)
Comment (hardy925) [reply serializable.ts]: keeping for now, we can chat about it before v2 release
